pypose/module/ipddp.py

```diff
@@ -0,0 +1,573 @@
+import time
+import torch as torch
+import torch.nn as nn
+from .. import bmv, bvmv, btdot
+
+class IPDDP(nn.Module):
+    r'''
+    Interior-point differential dynamic programming.
+
+    Args:
+        sys (:obj:`instance`): System dynamics of the optimal control problem.
+        stage_cost (:obj:`instance`): Stage cost of the optimal control problem.
+        terminal_cost (:obj:`instance`): Terminal cost of the optimal control problem.
+        cons (:obj:`instance`): Constraints of the optimal control problem.
+        n_cons (:obj:`int`): Dimension of constraints.
+        init_traj (:obj:`Dict`): Initial system trajectory.
+
+    A discrete-time system can be described as:
+
+    .. math::
+        \begin{align*}
+            \mathbf{x}_{t+1} &= \mathbf{f}(\mathbf{x}_{t}, \mathbf{u}_{t}, t)              \\
+        \end{align*}
+
+    where :math:`\mathbf{x}`, :math:`\mathbf{u}` are the state and input of the general nonlinear system.
+    The subscript :math:`\cdot_{t}` denotes the time step.
+
+    IPDDP finds the optimal nominal trajectory :math:`\mathbf{\tau}^*` =
+    :math:`\begin{Bmatrix} \mathbf{x}_t, \mathbf{u}_t \end{Bmatrix}_{0:T-1} \cup \begin{Bmatrix} \mathbf{x}_T \end{Bmatrix}`
+    for the following optimization problem:
+
+    .. math::
+        \begin{align*}
+          \mathbf{\tau}^* = \mathop{\arg\min}\limits_{\tau} & \sum\limits_{t=0}^{T-1}q(\mathbf{\tau}_t) + p(\mathbf{\tau}_T) \\
+          \mathrm{s.t.} \quad \mathbf{x}_0 &= \mathbf{x}_{\text{init}}, \\
+          \mathbf{x}_{t+1} &= \mathbf{f}(\mathbf{x}_{t}, \mathbf{u}_{t}, t),  \\
+          \mathbf{c}(\mathbf{\tau}_t) &\leq \mathbf{0}.
+        \end{align*}
+
+    where :math:`\mathbf{\tau}_t` = :math:`\begin{bmatrix} \mathbf{x}_t \\ \mathbf{u}_t
+    \end{bmatrix}`; :math:`q` and :math:`p` denote the stage and terminal costs, respectively;
+    :math:`\mathbf{c}` is the stage-wise inequality constraints. In addition to the notations
+    defined above, two additional variables :math:`\mathbf{s}` and :math:`\mathbf{y}` are
+    introduced, which denote the dual variable and the slack variable, respectively.
+
+    The IPDDP process can be summarised as iterative backward and forward recursions.
+
+    - The backward recursion.
```

Comment:
It seems two scenarios are implemented with one starts from a feasible trajectory and the other from an infeasible trajectory. But only one scenario's doc is provided? That's why there is no updated related to the slack variable y, although it's defined.

Reply:
fixed this.